dvc/data/stage.py

```diff
@@ -247,27 +250,88 @@ def _get_staging(odb: "ObjectDB") -> "ObjectDB":
     return ReferenceObjectDB(fs, fs_path, state=state)
 
 
-def _load_from_state(odb, staging, fs_path, fs, name):
+def _stage_raw_dir_hash(
+    staging: "ReferenceObjectDB",
+    tree: "Tree",
+    fs_path: "AnyFSPath",
+    fs: "FileSystem",
+    hash_info: "HashInfo",
+) -> "Tree":
+    assert tree.fs
+    staging.add(tree.fs_path, tree.fs, hash_info)
+    for key, _, oid in tree:
+        staging.add(fs.path.join(fs_path, *key), tree.fs, oid, verify=False)
```

Comment:
Btw, strictly speaking, we need to add the files first, so that at no point our odb is corrupted and so we avoid the risk of leaving it corrupted in a case of an error (not too relevant with memory odb, but still)